crates/base-db/src/input.rs

```diff
@@ -324,6 +328,97 @@ pub struct CrateData {
     pub channel: Option<ReleaseChannel>,
 }
 
+impl CrateData {
+    /// Check if [`other`] is almost equal to [`self`] ignoring `CrateOrigin` value.
+    pub fn eq_ignoring_origin(&self, other: &CrateData) -> bool {
+        // This method has some obscure bits. These are mostly there to be compliant with
+        // some patches. References to the patches are given.
+        if self.root_file_id != other.root_file_id {
+            return false;
+        }
+
+        if self.display_name != other.display_name {
+            return false;
+        }
+
+        if self.is_proc_macro != other.is_proc_macro {
+            return false;
+        }
+
+        if self.edition != other.edition {
+            return false;
+        }
+
+        if self.version != other.version {
+            return false;
+        }
+
+        let mut opts = self.cfg_options.clone();
+        opts.apply_diff(
+            CfgDiff::new(vec![], other.cfg_options.clone().into_iter().collect())
+                .expect("CfgOptions were expected to contain no duplicates."),
+        );
+
+        let mut cfgs = opts.into_iter();
+        if let Some(cfg) = cfgs.next() {
+            // Don't care if rust_analyzer CfgAtom is the only cfg in the difference set of self's and other's cfgs.
+            // https://github.com/rust-lang/rust-analyzer/blob/0840038f02daec6ba3238f05d8caa037d28701a0/crates/project-model/src/workspace.rs#L894
+            if cfgs.next().is_some() || cfg.to_string() != "rust_analyzer" {
+                return false;
+            }
+        }
```

Comment:
Deleaget https://doc.rust-lang.org/std/collections/struct.HashSet.html#method.difference to `CfgOptions::difference` on its field